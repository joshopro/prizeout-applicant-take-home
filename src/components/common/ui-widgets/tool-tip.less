.tool-tip {
    display: inline-block;
    position: relative;

    &.showing {
        article {
            display: block;
        }
    }

    article {
        background-color: #808997;
        border-radius: .75rem;
        color: #fff;
        display: none;
        font-size: .75rem;
        font-weight: normal;
        max-width: 18rem;
        min-width: 12rem;
        padding: .25rem .5rem;
        position: absolute;
        top: calc(100% + 1rem);
        left: -22%;
        line-height: 1rem;
    }

    header {
        cursor: pointer;
        display: inline-block;
    }
}

