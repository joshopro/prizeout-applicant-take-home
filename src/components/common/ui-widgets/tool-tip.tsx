import React, { useState } from 'react';
import Classnames from 'classnames';
import PropTypes from 'prop-types';
import { useHotkeys } from 'react-hotkeys-hook';

import './tool-tip.less';

interface ToolTipProps {
    content: string;
    id?: string;
    title: string;
}

export const ToolTip: React.FC<ToolTipProps> = ({ content, id, title }): React.ReactElement => {
    const [isShowing, setIsShowing] = useState(false),
        showToolTip = (): void => setIsShowing(true),
        hideToolTip = (): void => setIsShowing(false),
        classes: string = Classnames('tool-tip', { showing: isShowing }),
        elmId = id || `tt-id`,
        contentId = `${elmId}-content`;

    useHotkeys('esc', () => {
        if (document.activeElement === document.getElementById(elmId)) {
            hideToolTip();
        }
    });

    useHotkeys('enter', () => {
        if (document.activeElement === document.getElementById(elmId)) {
            showToolTip();
        }
    });

    return (
        <div
            onBlur={hideToolTip}
            onFocus={showToolTip}
            aria-hidden={!isShowing ? 'true' : 'false'}
            className={classes}
            id={elmId}
            tabIndex={0}
        >
            <header aria-describedby={contentId} onMouseEnter={showToolTip} onMouseLeave={hideToolTip}>
                {title}
            </header>
            <article id={contentId}>{content}</article>
        </div>
    );
};

ToolTip.propTypes = {
    content: PropTypes.string.isRequired,
    id: PropTypes.string,
    title: PropTypes.string.isRequired,
};
