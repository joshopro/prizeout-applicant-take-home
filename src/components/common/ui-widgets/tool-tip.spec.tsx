import { render, screen } from '../../../testing/test-utils';
import userEvent from '@testing-library/user-event';
import React from 'react';
import '@testing-library/jest-dom';

import { ToolTip } from './tool-tip';

describe('Test Tool Tip Component', () => {
    const content = `Tool Tip Content`;
    const title = `Tool Tip Title`;
    const id = 'fakeTip';

    test('Component matches snapshot', () => {
        const { asFragment } = render(<ToolTip content={content} id={id} title={title} />);
        expect(asFragment()).toMatchSnapshot();
    });

    test('Content is hidden by default', () => {
        render(<ToolTip content={content} id={id} title={title} />);

        // expect(screen.getByText(content)).not.toBeVisible();
    });

    test('Content shows when hovering on the header', () => {
        render(<ToolTip content={content} id={id} title={title} />);

        // expect(screen.getByText(content)).not.toBeVisible();
        userEvent.hover(screen.getByText(title));
        expect(screen.getByText(content)).toBeVisible();
    });

    test('Content hides when unhovering on the header', () => {
        render(<ToolTip content={content} id={id} title={title} />);

        // expect(screen.getByText(content)).not.toBeVisible();
        userEvent.hover(screen.getByText(title));
        expect(screen.getByText(content)).toBeVisible();
        userEvent.unhover(screen.getByText(title));
        // expect(screen.getByText(content)).not.toBeVisible();
    });
});
