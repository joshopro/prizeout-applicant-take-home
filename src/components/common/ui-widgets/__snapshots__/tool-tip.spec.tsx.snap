// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Tool Tip Component Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="tool-tip"
    id="fakeTip"
    tabindex="0"
  >
    <header
      aria-describedby="fakeTip-content"
    >
      Tool Tip Title
    </header>
    <article
      id="fakeTip-content"
    >
      Tool Tip Content
    </article>
  </div>
</DocumentFragment>
`;
